quill-jdbc-monix/src/main/scala/io/getquill/context/jdbc/AsyncJdbcContext.scala

```diff
@@ -0,0 +1,248 @@
+            withAutocommitBracket(conn, conn => {
+              wrap(conn).flatMap { conn =>
+                currentConnection.update(Some(conn))
+                f.onCancelRaiseError(new IllegalStateException(
+                  "The task was cancelled in the middle of a transaction."
+                )).doOnFinish {
+                  case Some(error) =>
+                    conn.rollback()
+                    Task.raiseError(error)
+                  case None =>
+                    wrap(conn.commit())
+                }
+              }
+            })
+          } { conn =>
+            wrap(currentConnection.update(None))
+          }
+      }
+    } yield result
+
+    boundary {
+      schedule(dbEffects)
+        .executeWithOptions(_.enableLocalContextPropagation)
+    }
+  }
+
+  // Override with sync implementation so will actually be able to do it.
+  override def probe(sql: String): Try[_] = Try {
+    val c = dataSource.getConnection
+    try {
+      c.createStatement().execute(sql)
+    } finally {
+      c.close()
+    }
+  }
+
+  /**
+   * In order to allow a ResultSet to be consumed by an Observeable, a ResultSet iterator must be created.
+   * Since Quill provides a extractor for an individual ResultSet row, a single row can easily be cached
+   * in memory. This allows for a straightforward implementation of a hasNext method.
+   */
+  class ResultSetIterator[T](rs: ResultSet, extractor: Extractor[T]) extends BufferedIterator[T] {
+
+    private[this] var state = 0 // 0: no data, 1: cached, 2: finished
+    private[this] var cached: T = null.asInstanceOf[T]
+
+    protected[this] final def finished(): T = {
+      state = 2
+      null.asInstanceOf[T]
+    }
+
+    /** Return a new value or call finished() */
+    protected def fetchNext(): T =
+      if (rs.next()) extractor(rs)
+      else finished()
+
+    def head: T = {
+      prefetchIfNeeded()
+      if (state == 1) cached
+      else throw new NoSuchElementException("head on empty iterator")
+    }
+
+    private def prefetchIfNeeded(): Unit = {
+      if (state == 0) {
+        cached = fetchNext()
+        if (state == 0) state = 1
+      }
+    }
+
+    def hasNext: Boolean = {
+      prefetchIfNeeded()
+      state == 1
+    }
+
+    def next(): T = {
+      prefetchIfNeeded()
+      if (state == 1) {
+        state = 0
+        cached
+      } else throw new NoSuchElementException("next on empty iterator");
+    }
+  }
+
+  /**
+   * Override to enable specific vendor options needed for streaming
+   */
+  protected def prepareStatementForStreaming(sql: String, conn: Connection, fetchSize: Option[Int]) = {
+    val stmt = conn.prepareStatement(sql, ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY)
+    fetchSize.foreach { size =>
+      stmt.setFetchSize(size)
+    }
+    stmt
+  }
+
+  def streamQuery[T](fetchSize: Option[Int], sql: String, prepare: Prepare = identityPrepare, extractor: Extractor[T] = identityExtractor): Observable[T] =
```

Comment:
I think Monix has support for cancellations, should we implement it?

Reply:
Do you mean have a method that returns a composite `Cancellable` object? What are some good use-cases and/or advantages of that? I think both `Task` and `Observeable` return a cancellable future when run.

Reply:
nvm, `bracket` already takes care of cancellations: https://github.com/monix/monix/blob/master/monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/ConcatMapObservable.scala#L42